File: 065.rb
# 065
# Valid Number
# 2015-12-20
#####################################################
# @param {String} s
# @return {Boolean}
def is_number(s)
    i = 0
    i+=1 while s[i] == ' '
    return false if i == s.size
    i+=1 if s[i] == '+' or s[i] == '-'
    i+=1 while myisdigit(s, i)
    if s[i] == '.'
        return false if not myisdigit(s, i-1) and not myisdigit(s, i+1)
        i+=1
    end
    i+=1 while myisdigit(s, i)
    if s[i] == 'e'
        return false if not myisdigit(s, i-1) and not myisdigit(s, i-2)
        return false if not myisdigit(s, i+1) and not myisdigit(s, i+2)
        i+=1
        i+=1 if (s[i] == '+' or s[i] == '-') and myisdigit(s, i+1)
    end
    i+=1 while myisdigit(s, i)
    i+=1 while s[i] == ' '
    i == s.size
end

def myisdigit(s, i)
  return false if i < 0 or i >= s.size
  Float(s[i]) != nil rescue false
end

#####################################################
s = " 1234 "
r = is_number(s)
puts r

File: 065.js
//  065
//  Valid Number
//  2015-12-20
/////////////////////////////////////////////////////
/**
 * @param {string} s
 * @return {boolean}
 */
var isNumber = function(s) {
    var i = 0;
    var l = s.length;
    while (i < l && s[i] == ' ')
        i++;
    if (i == l)
        return false;
    if (s[i] == '+' || s[i] == '-')
        i++;
    while (myisdigit(s, i))
        i++;
    if (i < l && s[i] == '.') {
        if (!myisdigit(s, i-1) && !myisdigit(s, i+1))
            return false;
        i++;
    }
    while (myisdigit(s, i))
        i++;
    if (i < l && s[i] == 'e') {
        if (!myisdigit(s, i-1) && !myisdigit(s, i-2))
            return false;
        if (!myisdigit(s, i+1) && !myisdigit(s, i+2))
            return false;
        i++;
        if ((s[i] == '+' || s[i] == '-') && myisdigit(s, i+1))
            i++;
    }
    while (myisdigit(s, i))
        i++;
    while (i < l && s[i] == ' ')
        i++;
    return i == l;
};

function myisdigit(s, i) {
    if (i < 0 || i >= s.length || s[i] == ' ')
        return false;
    return !isNaN(s[i]);
}

/////////////////////////////////////////////////////
var s = " .";
var r = isNumber(s);
console.log(r);

